src/gil.rs

```diff
@@ -233,7 +248,12 @@ impl GILGuard {
 
         unsafe {
             let gstate = ffi::PyGILState_Ensure(); // acquire GIL
+            GIL_COUNT.with(|c| c.set(c.get() + 1));
+
+            // Release objects that were dropped since last GIL acquisition
             let pool: &'static mut ReleasePool = &mut *POOL;
+            pool.release_pointers();
```

Comment:
Why do we need to release pointers here?
To reduce memory usage?

Reply:
Two reasons:

1. Yes, memory usage - we're now releasing the pointers sooner
2. This (https://github.com/PyO3/pyo3/issues/679#issuecomment-562964486) suggested to me that it's important in the long run we migrate to releasing objects on GIL acquire, not GIL exit.

Reply:
Ah, I understand. Thank you for the reference.